### PR TITLE
71 shadow runtimes

### PR DESCRIPTION
**Pull request by tafseer-khan**
Tests: https://asciinema.org/a/cHr4a4Z4kr1TrmlOVJzwCeNfZ

Benchmark Test: https://asciinema.org/a/6r50DLr9tETp18F81TBHplOmX

### CODE REVIEW

File: protocols/substrate/components/database/database/smartops.go [hunk 1]
```diff
@@ -28,6 +27,6 @@ func (f *Database) Application() string {
 	return f.dbContext.ApplicationId
 }
 
-func (f *Database) Project() (cid.Cid, error) {
-	return cid.Decode(f.dbContext.ProjectId)
+func (f *Database) Project() string {
```
Comment (samyfodil) [inline protocols/substrate/components/database/database/smartops.go:30]: why not a cid.Cid

File: protocols/substrate/components/http/api.go [hunk 2]
```diff
@@ -23,9 +23,13 @@ func (s *Service) handle(w goHttp.ResponseWriter, r *goHttp.Request) error {
...
 	}
 
+	if len(pickServiceables) != 1 {
+		return fmt.Errorf("lookup returned %d serviceables, expected 1", len(pickServiceables))
+	}
+
 	pick, ok := pickServiceables[0].(iface.Serviceable)
 	if !ok {
-		return fmt.Errorf("matched serviceable is not a http serviceable")
+		return fmt.Errorf("matched serviceable is not a http serviceable %#v", pickServiceables[0])
```
Comment (samyfodil) [inline protocols/substrate/components/http/api.go:32]: implement String() for serviceable & avoid %#v

File: vm/call.go [hunk 3]
```diff
@@ -1,75 +1,36 @@
...
-		}
-
-		return source, nil
-	}
-}
-
 // Call takes instance and id, then calls the moduled function. Returns an error.
-func (f *FunctionInstance) Call(runtime vm.Runtime, id interface{}) error {
-	moduleName, err := f.moduleName()
+func (w *WasmModule) Call(runtime vm.Runtime, id interface{}) error {
```
Comment (samyfodil) [inline vm/call.go:12]: func (w *WasmModule) Call(runtime vm.Runtime, id uint32) error {
...

ret := fx.RawCall(ctx, uint64(id))

Comment (tafseer-khan) [reply vm/call.go:12]: d *Dfunc

File: vm/cache/cache.go [hunk 4]
```diff
@@ -32,52 +25,72 @@ func (c *Cache) Close() {
 // The serviceable itself is stored by the id of the serviceable.
 func New() *Cache {
 	return &Cache{
-		cacheMap: make(map[string]map[string]iface.Serviceable, 0),
+		cacheMap: make(map[string]map[string]cacheItem, 0),
```
Comment (samyfodil) [inline vm/cache/cache.go:28]: asset cid should be stored in servisable. so just use serviceable

**Comment by tafseer-khan** [comment]
AssetID should be interface method, dont store in cache item

File: protocols/substrate/components/http/function/new.go [hunk 5]
```diff
@@ -7,6 +7,7 @@ import (
 	commonIface "github.com/taubyte/go-interfaces/services/substrate/components"
 	iface "github.com/taubyte/go-interfaces/services/substrate/components/http"
 	"github.com/taubyte/go-interfaces/services/tns"
+	spec "github.com/taubyte/go-specs/common"
 	"github.com/taubyte/go-specs/extract"
 	"github.com/taubyte/tau/protocols/substrate/components/http/common"
 	tvm "github.com/taubyte/tau/vm"
```
Comment (samyfodil) [inline protocols/substrate/components/http/function/new.go:13]: why call it tvm. just refer to it as vm

File: protocols/substrate/components/http/function/types.go [hunk 6]
```diff
@@ -21,15 +18,15 @@ type Function struct {
...
-
 	readyDone  bool
 	readyCtx   context.Context
 	readyCtxC  context.CancelFunc
 	readyError error
 
 	instanceCtx  context.Context
 	instanceCtxC context.CancelFunc
+
+	module *tvm.WasmModule
```
Comment (samyfodil) [inline protocols/substrate/components/http/function/types.go:29]: WasmModule -> DFunc

File: protocols/substrate/components/http/lookup_test.go [hunk 7]
```diff
@@ -116,8 +116,8 @@ func checkFunction(s *Service, id string, matcher commonIface.MatchDefinition) e
 		return errors.New("Not ok")
 	}
 
-	if reflect.DeepEqual(functions[id], *f.Config()) == false {
-		return fmt.Errorf("Expected: %#v, got: %#v", functions[id], f.Config())
+	if reflect.DeepEqual(functions[id], *f.Structure()) == false {
```
Comment (samyfodil) [inline protocols/substrate/components/http/lookup_test.go:119]: keep it to Config

File: protocols/substrate/components/pubsub/function/new.go [hunk 8]
```diff
@@ -29,19 +31,26 @@ func New(srv iface.Service, mmi common.MessagingMapItem, config structureSpec.Fu
 		f.readyCtxC()
 	}()
 
-	_f, err := srv.Cache().Add(f)
+	if f.dFunc, err = vm.New(f.instanceCtx, f, spec.DefaultBranch, f.commit); err != nil {
+		return nil, fmt.Errorf("initializing vm module failed with: %w", err)
+	}
+
+	f.assetId, err = cache.ComputeServiceableCid(f, spec.DefaultBranch)
```
Comment (samyfodil) [inline protocols/substrate/components/pubsub/function/new.go:38]: call ComputeServiceableCid -> ResolveServiceableCid

File: protocols/substrate/components/p2p/function/function.go [hunk 9]
```diff
@@ -19,32 +17,24 @@ func (f *Function) Commit() string {
...
-	instance, runtime, plugin, err := f.function.Instantiate(
-		commonIface.FunctionContext{
-			Config:      f.config,
-			Project:     f.matcher.Project,
-			Application: f.matcher.Application,
-		},
-		spec.DefaultBranch,
-		f.commit,
-	)
+	runtime, pluginApi, err := f.Function.Instantiate()
```
Comment (samyfodil) [inline protocols/substrate/components/p2p/function/function.go:29]: f.Instantiate()

File: protocols/substrate/components/p2p/function/function.go [hunk 10]
```diff
@@ -61,7 +51,7 @@ func (f *Function) Handle(cmd *command.Command) (t time.Time, res response.Respo
 		}
 	}
 
-	return time.Now(), res, instance.Call(runtime, ev.Id)
+	return time.Now(), res, f.Function.Call(runtime, ev.Id)
```
Comment (samyfodil) [inline protocols/substrate/components/p2p/function/function.go:54]: f.Call(...)

File: protocols/substrate/components/pubsub/function/function.go [hunk 11]
```diff
@@ -20,28 +18,20 @@ func (f *Function) Commit() string {
...
-	instance, runtime, plugin, err := f.function.Instantiate(
-		components.FunctionContext{
-			Config:      f.config,
-			Project:     f.matcher.Project,
-			Application: f.matcher.Application,
-		},
-		spec.DefaultBranch,
-		f.commit,
-	)
+	runtime, pluginApi, err := f.Function.Instantiate()
```
Comment (samyfodil) [inline protocols/substrate/components/pubsub/function/function.go:26]: f.Instantiate()

File: protocols/substrate/components/pubsub/function/function.go [hunk 12]
```diff
@@ -53,7 +43,7 @@ func (f *Function) HandleMessage(msg *pubsub.Message) (t time.Time, err error) {
 		return t, fmt.Errorf("exited: %d", val)
 	}
 
-	return time.Now(), instance.Call(runtime, ev.Id)
+	return time.Now(), f.Function.Call(runtime, ev.Id)
```
Comment (samyfodil) [inline protocols/substrate/components/pubsub/function/function.go:46]: f.Call(...)

File: protocols/substrate/components/http/function/new.go [hunk 13]
```diff
@@ -40,15 +42,22 @@ func New(srv iface.Service, object tns.Object, matcher *common.MatchDefinition)
 		f.readyCtxC()
 	}()
 
-	_f, err := srv.Cache().Add(f)
+	f.assetId, err = cache.ComputeServiceableCid(f, f.branch)
```
Comment (samyfodil) [inline protocols/substrate/components/http/function/new.go:45]: ComputeServiceableCid -> ResolveServiceableCid

**Review by samyfodil — CHANGES_REQUESTED** [review]

**Review by samyfodil — APPROVED** [review]

**Pull request merged**